rosbag2_compression/src/rosbag2_compression/zstd_compressor.cpp

```diff
@@ -0,0 +1,136 @@
+
+#include "logging.hpp"
+#include "rosbag2_compression/zstd_compressor.hpp"
+
+namespace
+{
+// Increasing the compression level will:
+//   - Increase the time taken to compress
+//   - Decrease the size of the compressed data
+// Setting to zero uses Zstd's default value of 3.
+constexpr const int DEFAULT_ZSTD_COMPRESSION_LEVEL = 1;
+
+constexpr const char COMPRESSION_IDENTIFIER[] = "zstd";
+
+std::vector<uint8_t> get_input_buffer(const std::string & uri)
+{
+  // Get the file size
+  const auto decompressed_buffer_length = rosbag2_storage::FilesystemHelper::get_file_size(uri);
+  // Read in buffer, handling accordingly
+  auto file_pointer = std::fopen(uri.c_str(), "rb");
+  if (file_pointer == nullptr)
+  {
+    fclose(file_pointer);
+    throw std::runtime_error("Error opening file");
+  }
+  // Allocate and read in
+  std::vector<uint8_t> decompressed_buffer;
+  decompressed_buffer.reserve(decompressed_buffer_length);
+  fread(decompressed_buffer.data(), sizeof(uint8_t), decompressed_buffer_length, file_pointer);
+  if (ferror(file_pointer))
+  {
+    fclose(file_pointer);
+    throw std::runtime_error("Unable to read file");
+  }
+  fclose(file_pointer);
+  return decompressed_buffer;
+}
+
+void write_output_buffer(
+  const std::vector<uint8_t> & output_buffer,
+  const std::string & uri)
+{
+  auto file_pointer = std::fopen(uri.c_str(), "wb");
+  fwrite(output_buffer.data(), sizeof(uint8_t), output_buffer.capacity(), file_pointer);
+  if (ferror(file_pointer))
+  {
+    fclose(file_pointer);
+    throw std::runtime_error("Unable to write compressed file");
+  }
+  fclose(file_pointer);
+}
+
+void throw_on_zstd_error(const size_t compression_result)
+{
+  if (ZSTD_isError(compression_result)) {
+    std::stringstream error;
+    error << "ZSTD compression error: " << ZSTD_getErrorName(compression_result);
+    throw std::runtime_error(error.str());
+  }
+  ROSBAG2_COMPRESSION_LOG_DEBUG("ZSTD compressed file.");
+}
+
+void print_compression_statistics(
+  std::chrono::high_resolution_clock::time_point start,
+  std::chrono::high_resolution_clock::time_point end,
+  size_t decompressed_size, size_t compressed_size)
+{
+  const auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
+  const auto compression_ratio =
+    static_cast<double>(decompressed_size) / static_cast<double>(compressed_size);
+  ROSBAG2_COMPRESSION_LOG_DEBUG_STREAM(
+    "Compression statistics:\n" <<
+      "Time: " << duration.count() << " microseconds" <<
+      "Compression Ratio: " << compression_ratio
+  );
+}
+}  // namespace
+
+namespace rosbag2_compression
+{
+
+std::string ZstdCompressor::compress_uri(const std::string & uri)
+{
+  const auto start = std::chrono::high_resolution_clock::now();
+  const auto compressed_uri = uri + "." + get_compression_identifier();
+  const auto decompressed_buffer = get_input_buffer(uri);
+  const auto decompressed_buffer_length = decompressed_buffer.capacity();
```

Comment:
why capacity and not size?

Reply:
`capacity` returns the size allocated by `reserve`
http://www.cplusplus.com/reference/vector/vector/capacity/

Reply:
What I mean here is that `capacity` can technically be larger than it's actual `size`.  I think that comment also applies for https://github.com/ros2/rosbag2/pull/220/files#diff-db7307286bf30f6efe78352f50c0b53bR63